content/content-tool/tool/src/java/org/sakaiproject/content/entityproviders/ContentEntityProvider.java

```diff
@@ -260,6 +278,151 @@ public String getHtmlForRef(EntityView view, Map<String, Object> params) throws
+	 *
+	 */
+	@EntityCustomAction(action = "direct-upload", viewKey = EntityView.VIEW_NEW)
+	public String uploadFileToSite(EntityView view,  Map<String, Object> params) {
+
+		if (!serverConfigurationService.getBoolean("content.direct.upload.enabled", Boolean.FALSE)) {
+			throw new SecurityException("The direct-upload service is not enabled for your instance.");
+		}
+
+		User currentUser = userDirectoryService.getCurrentUser();
+		if (currentUser == null || StringUtils.isBlank(currentUser.getId())) {
+			throw new SecurityException("You must be logged in to use the direct-upload service.");
+		}
+		String context = (String) params.get("context");
+
+		String collectionId = contentHostingService.getSiteCollection(context);
+
+		String uploadFolderName = "";
+
+		if (contentHostingService.allowAddCollection(collectionId)) {
+			// If user has content.new permission add file to instructor folder
+			uploadFolderName = contentHostingService.getInstructorUploadFolderName();
+		} else if (contentHostingService.allowGetCollection(collectionId)) {
+			// If user has content.read permission add file to student folder
+			uploadFolderName = contentHostingService.getStudentUploadFolderName();
+		} else {
+			// If user does not have either permission in the collection throw a security exception
+			throw new SecurityException("You must have at least read access for resources in the site to use direct-upload.");
+		}
+
+		String errorMessage;
+
+		try {
+			ContentCollection uploadsFolder = getDirectUploadsFolder(collectionId, uploadFolderName);
+
+			if (uploadsFolder == null) {
+				errorMessage = "Unable to get uploads folder.";
+			} else {
+
+				DiskFileItem fileItem = (DiskFileItem) params.get("upload");
+
+				String[] fileNameParts = fileItem.getName().split("\\.(?=[^.]+$)");
+
+				String basename = StringUtils.equals(uploadFolderName, contentHostingService.getStudentUploadFolderName()) ? currentUser.getDisplayId() + "_" + fileNameParts[0] : fileNameParts[0];
+
+				securityService.pushAdvisor(allowedAdvisor);
+
+				ContentResourceEdit resourceEdit = contentHostingService.addResource(
+						uploadsFolder.getId(),
+						basename,
+						fileNameParts[1],
+						ContentHostingService.MAXIMUM_ATTEMPTS_FOR_UNIQUENESS);
+
+				resourceEdit.setContent(fileItem.getInputStream());
+				resourceEdit.setContentType(fileItem.getContentType());
+				resourceEdit.setContentLength(fileItem.getSize());
+
+				contentHostingService.commitResource(resourceEdit, NotificationService.NOTI_NONE);
+				JSONObject success = new JSONObject();
+				success.put("uploaded", 1);
+				success.put("fileName", fileItem.getName());
+				success.put("url", resourceEdit.getUrl());
+
+				return success.toJSONString();
+			}
+
+		} catch (TypeException | InconsistentException | IdInvalidException | IdUnusedException e) {
+			log.error("Exception when trying to get the direct upload folder for site " + context + ".", e);
```

Comment:
log.error("Exception when trying to get the direct upload folder for site {}.", context, e);